src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/MessageBody.cs

```diff
@@ -21,80 +23,199 @@ protected MessageBody(Frame context)
             {
-                send100Continue = Interlocked.Exchange(ref _send100Continue, 0);
+                TryProduceContinue();
+
+                // Incomplete Task await result
+                return ReadAsyncAwaited(task, buffer);
             }
-            if (send100Continue == 1)
+            else
             {
-                _context.FrameControl.ProduceContinue();
+                var readSegment = task.Result;
+                var consumed = CopyReadSegment(readSegment, buffer);
+
+                return consumed == 0 ? TaskCache<int>.DefaultCompletedTask : Task.FromResult(consumed);
             }
-            return result;
         }
 
-        public Task Consume(CancellationToken cancellationToken = default(CancellationToken))
+        private async Task<int> ReadAsyncAwaited(ValueTask<ArraySegment<byte>> currentTask, ArraySegment<byte> buffer)
+        {
+            return CopyReadSegment(await currentTask, buffer);
+        }
+
+        private int CopyReadSegment(ArraySegment<byte> readSegment, ArraySegment<byte> buffer)
         {
-            ValueTask<int> result;
-            var send100checked = false;
-            do
+            var consumed = Math.Min(readSegment.Count, buffer.Count);
+
+            if (consumed != 0)
+            {
+                Buffer.BlockCopy(readSegment.Array, readSegment.Offset, buffer.Array, buffer.Offset, consumed);
+                ConsumedBytes(consumed);
+            }
+
+            return consumed;
+        }
+
+        public Task CopyToAsync(Stream destination, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var peekTask = PeekAsync(cancellationToken);
+
+            while (peekTask.IsCompleted)
             {
-                result = ReadAsyncImplementation(default(ArraySegment<byte>), cancellationToken);
-                if (!result.IsCompleted)
-                {
-                    if (!send100checked)
-                    {
-                        if (Interlocked.Exchange(ref _send100Continue, 0) == 1)
-                        {
-                            _context.FrameControl.ProduceContinue();
-                        }
-                        send100checked = true;
-                    }
-                    // Incomplete Task await result
-                    return ConsumeAwaited(result.AsTask(), cancellationToken);
-                }
                 // ValueTask uses .GetAwaiter().GetResult() if necessary
-                else if (result.Result == 0)
+                var segment = peekTask.Result;
+
+                if (segment.Count == 0)
                 {
-                    // Completed Task, end of stream
                     return TaskCache.CompletedTask;
                 }
 
-            } while (true);
+                Task destinationTask;
+                try
+                {
+                    destinationTask = destination.WriteAsync(segment.Array, segment.Offset, segment.Count, cancellationToken);
```

Comment:
You should add a test that verifies this call. Call `CopyToAsync` on some mock stream and check that the address of the array is the same as what `SocketInput.PeekAsync()` gives you.